apps/prairielearn/src/server.js

```diff
@@ -590,6 +590,15 @@ module.exports.initExpress = function () {
     require('./pages/instructorRequestCourse/instructorRequestCourse.js'),
   ]);
 
+  app.use('/pl/navbar/course/:course_id/switcher', [
+    require('./middlewares/authzCourseOrInstance'),
```

Comment:
Is this middleware actually needed? As mentioned in a comment, this is blocking a user without course access (but with course instance access) from seeing the list of courses they have access to. I believe this should be either

1. a path that supports an optional course_instance_id, or 
2. get rid of this middleware, and assume no overrides. Note that the dropdown is already disabled in case of overrides, so this would only affect users that explicitly use the URL, and in any case would only select courses they have authorized access to. Also, with no overrides `course_id` only affects which dropdown entry is active, and can be obtained from the query params.

Reply:
You could even change the param name to `unsafe_course_id` to make this clear in the page code.

Reply:
I was typing out this response at the top level before you left this comment, but I'll shift it here to keep things threaded.

I actually observe subtly different behavior than what you describe. If I have no course access but I do have student data access, the course instance switcher works fine, but the _course_ switcher 403s. Can you triple check that the course _instance_ switcher is what was failing for you, and if so, give precise steps to reproduce that?

I think my investigation of this is revealing some inconsistent, if unexpected or downright broken, behavior with auth. If I have `None` course permissions, I can actually still access the course admin pages through the course instance routes. For instance, consider a course 1 with a course instance 2. `/pl/course/1/course_admin` will 403, but `/pl/course_instance/2/instructor/course_admin` does not, and the latter shows the same content as the first route.

I think your suggestions are better fixes, but let's still talk about this, even if we fix it in another PR. We should do one of the following:

- Forbid access to `course_admin` pages if the user does not have a course role > `None`
- Allow access to `/pl/.../:course_id/...` if the user has *any* permissions in the course instance

To be honest, I'm not sure entirely what `None` is supposed to represent in course permissions. That role is described as "Cannot see any course content" on the staff page, but the current behavior certainly allows access to it. And even then, if someone has a `None` course role but some student data access, they can access all course content through the course instance questions page!

Reply:
To respond to your specific suggestions: yeah, I think we can remove the authz middleware entirely here. As you point out, we won't actually leak any data even without that, and we don't even fetch the switcher contents when overrides are active.

Reply:
> I actually observe subtly different behavior than what you describe. If I have no course access but I do have student data access, the course instance switcher works fine, but the _course_ switcher 403s. Can you triple check that the course _instance_ switcher is what was failing for you, and if so, give precise steps to reproduce that?

Indeed it is the course switcher that fails in the latest version. I believe that comment is from a previous version which seems to be fixed now, or I got mixed up.

> I think your suggestions are better fixes, but let's still talk about this, even if we fix it in another PR. We should do one of the following:
> 
> * Forbid access to `course_admin` pages if the user does not have a course role > `None`
> * Allow access to `/pl/.../:course_id/...` if the user has _any_ permissions in the course instance
>
> To be honest, I'm not sure entirely what `None` is supposed to represent in course permissions. That role is described as "Cannot see any course content" on the staff page, but the current behavior certainly allows access to it. And even then, if someone has a `None` course role but some student data access, they can access all course content through the course instance questions page!

This has some relationship with #8850. I agree that the first option makes sense and would probably fix that issue. That said, this should probably be done in a dedicated PR, since there are corner cases to consider, for example: would a TA with course instance access but no course access be able to open the question preview page for a question that is in an assessment they have access to?

* If yes, how do we filter this?
* If no, we should probably also remove links that point to the preview from an assessment context ("instructor view" links in an assessment instance log, the instructor box in an instance question, issues page, etc.).

Reply:
I think we can move the braoder discussion of this to #8850.

I removed the authz middleware and made the other necessary changes in 658ecb5. Note that I switched from `res.locals.user.user_id` to `res.locals.authn_user.user_id` since the former isn't populated without the middleware. I believe this is safe, as the only time the `authn_user` ID might not be the right choice is if an auth override is in place, but if that's the case, we're already not rendering the dropdown anyways.

Reply:
Yup, looks good. Resolving.